compiler/rustc_hir_analysis/src/check/region.rs

```diff
@@ -885,3 +892,190 @@ pub fn region_scope_tree(tcx: TyCtxt<'_>, def_id: DefId) -> &ScopeTree {
 
     tcx.arena.alloc(scope_tree)
 }
+
+/// Check for shared or mutable references of `static mut`
+pub fn maybe_static_mut_ref(
+    tcx: TyCtxt<'_>,
+    block: Option<Block<'_>>,
+    f: Option<(FnKind<'_>, hir::BodyId)>,
+) {
+    fn maybe_expr_static_mut(
+        tcx: TyCtxt<'_>,
+        expr: hir::Expr<'_>,
+        span: Span,
+        e2024: bool,
+        safe: bool,
+        hir_id: HirId,
+    ) {
+        if let hir::ExprKind::AddrOf(borrow_kind, m, expr) = expr.kind
+            && matches!(borrow_kind, BorrowKind::Ref)
+            && let hir::ExprKind::Path(qpath) = expr.kind
+            && let hir::QPath::Resolved(_, path) = qpath
+            && let hir::def::Res::Def(def_kind, _) = path.res
+            && let hir::def::DefKind::Static(mt) = def_kind
+            && matches!(mt, Mutability::Mut)
+        {
+            handle_static_mut_ref(
+                tcx,
+                span,
+                qpath_to_string(&qpath),
+                e2024,
+                matches!(m, Mutability::Mut),
+                safe,
+                hir_id,
+            );
+        }
+    }
+
+    fn check_stmts(tcx: TyCtxt<'_>, hir_id: HirId, stmts: &[Stmt<'_>], safe: bool) {
```

Comment:
This seems a little bit more manual than I would expect, could you use a HIR `Visitor` here instead?

Reply:
yeah I've suggested similar [above](https://github.com/rust-lang/rust/pull/117556#issuecomment-1837169227).